crates/sui-types/src/multisig_legacy.rs

```diff
@@ -0,0 +1,364 @@
+};
+use once_cell::sync::OnceCell;
+use roaring::RoaringBitmap;
+use schemars::JsonSchema;
+use serde::{ser::SerializeSeq, Deserialize, Deserializer, Serialize, Serializer};
+use serde_with::serde_as;
+use shared_crypto::intent::IntentMessage;
+use std::hash::{Hash, Hasher};
+
+use crate::{
+    base_types::SuiAddress,
+    crypto::{PublicKey, Signature},
+    error::SuiError,
+};
+
+pub type WeightUnit = u8;
+pub type ThresholdUnit = u16;
+pub const MAX_SIGNER_IN_MULTISIG: usize = 10;
+
+/// The struct that contains signatures and public keys necessary for authenticating a MultiSig.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+pub struct MultiSigLegacy {
+    /// The plain signature encoded with signature scheme.
+    sigs: Vec<CompressedSignature>,
+    /// A bitmap that indicates the position of which public key the signature should be authenticated with.
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "SuiBitmap")]
+    bitmap: RoaringBitmap,
+    /// The public key encoded with each public key with its signature scheme used along with the corresponding weight.
+    multisig_pk: MultiSigPublicKeyLegacy,
+    /// A bytes representation of [struct MultiSigLegacy]. This helps with implementing [trait AsRef<[u8]>].
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+/// This initialize the underlying bytes representation of MultiSig. It encodes
+/// [struct MultiSigLegacy] as the MultiSig flag (0x03) concat with the bcs bytes
+/// of [struct MultiSigLegacy] i.e. `flag || bcs_bytes(MultiSig)`.
+impl AsRef<[u8]> for MultiSigLegacy {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| {
+                let as_bytes = bcs::to_bytes(self).expect("BCS serialization should not fail");
+                let mut bytes = Vec::with_capacity(1 + as_bytes.len());
+                bytes.push(SignatureScheme::MultiSig.flag());
+                bytes.extend_from_slice(as_bytes.as_slice());
+                Ok(bytes)
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for MultiSigLegacy {
+    fn eq(&self, other: &Self) -> bool {
+        self.sigs == other.sigs
+            && self.bitmap == other.bitmap
+            && self.multisig_pk == other.multisig_pk
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for MultiSigLegacy {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for MultiSigLegacy {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for MultiSigLegacy {
+    fn verify_secure_generic<T>(
+        &self,
+        value: &IntentMessage<T>,
+        author: SuiAddress,
+        _aux_verify_data: AuxVerifyData,
+    ) -> Result<(), SuiError>
+    where
+        T: Serialize,
+    {
+        if self.multisig_pk.pk_map.len() > MAX_SIGNER_IN_MULTISIG {
```

Comment:
can we convert here the object to MultiSig and use its verify_secure_generic?
(and a similar approach for some of the large functions below)